trieste/ask_tell_optimization.py

```diff
@@ -303,7 +303,7 @@ def __repr__(self) -> str:
     @classmethod
     def from_record(
```

Comment:
shouldn't we modify `to_record` or `to_result` method to accept a path and can save the results?

Reply:
also, with those two methods we have a serious issue for gpflux and keras, we cannot obtain results with ask-tell interface, right?

Reply:
1. `to_record` returns a `Record` which can be saved simply by calling `save` on it. Similarly `to_result`.
2. I don't understand the question. Ask tell works exactly like it did before. We already weren't able to call `to_record` or `to_result` with gpflux/keras as it requires deepcopying the model. And saving to disk isn't enabled by default anyway. But an ask tell loop can still be executed.

Reply:
Hoping to take another look at deepcopying keras models soon.

Reply:
1) OK, perhaps adding a sentence or two about saving output from `to_result` in the ask-tell notebook?
2) I wasn't aware it wasn't possible. With `BayesianOptimizer`, you can still obtain `OptimizationResult` with gpflux or keras model even if they cannot use state tracking, while here we cannot obtain `OptimizationResult` at all - why is deepcopying the models necessary here but not in the `BayesianOptimizer` case?

Reply:
`BayesianOptimizer` provides an `OptimizationResult` containing the final model state after the loop is finished. If `AskTellOptimizer` returned the uncopied model state then it would be modified as soon as you run another `tell`.

It might be useful to have a models property so you can more cleanly inspect the models during optimization (so `ask_tell.models` rather than `ask_tell._models`) but that's definitely another PR.

Reply:
`models` property is a good idea (another that I would find useful is `num_query_points` in acq rules) - I agree, a separate PR for these...

Reply:
I'm not sure what is the solution but its still a problem that we cannot get `OptimizationResult` if we use a model that doesn't have deepcopy method

I think it might be fine for trieste, eventually we will sort out deepcopying for the models we support, but for any researcher using trieste and implementing their own models its unlikely they will sort out deepcopying as well and its not good if they cannot get basic things like optimization result

perhaps we could make model an optional return in the `OptimizationResult`? combined with model property and some caveats in the docs that might be good enough

Reply:
(We could even add a `copy: bool = True` argument to `to_result`, so people can get an optimization result that uses the non-copied model/data/state if they want.)